src/gateway/sharding/shard_manager.rs

```diff
@@ -214,169 +195,177 @@ impl ShardManager {
     /// use serenity::model::id::ShardId;
     /// use serenity::prelude::*;
     ///
-    /// # async fn run(client: Client) {
+    /// # async fn run(mut client: Client) {
     /// // restart shard ID 7
     /// client.shard_manager.restart(ShardId(7)).await;
     /// # }
     /// ```
     ///
     /// [`ShardRunner`]: super::ShardRunner
     #[cfg_attr(feature = "tracing_instrument", instrument(skip(self)))]
-    pub async fn restart(&self, shard_id: ShardId) {
+    pub async fn restart(&mut self, shard_id: ShardId) {
         info!("Restarting shard {shard_id}");
-        self.shutdown(shard_id, 4000).await;
-        self.boot(shard_id, false);
-    }
-
-    /// Returns the [`ShardId`]s of the shards that have been instantiated and currently have a
-    /// valid [`ShardRunner`].
-    ///
-    /// [`ShardRunner`]: super::ShardRunner
-    #[cfg_attr(feature = "tracing_instrument", instrument(skip(self)))]
-    pub async fn shards_instantiated(&self) -> Vec<ShardId> {
-        self.runners.lock().await.keys().copied().collect()
+        self.shutdown(shard_id, 4000);
+        self.boot(shard_id, false).await;
```

Comment:
This is no longer waiting for the shard to disconnect before restarting, as `shutdown` is now just firing off a message, which worries me quite a bit as a possible race condition. The `boot` call being async is also misleading as this code path is interchangably using unbounded and bounded senders for reasons?

Can you implement the waiting for shutdown, by using something like tokio::Notify (or a simpler solution: sending a oneshot zst channel over)?

Reply:
I'm actually having a hard time figuring out how the manager was waiting for the shutdown to finish before. Do you know what it was doing?

Reply:
Oh, seems like the runner was calling `self.manager.shutdown_finished()` which would make the manager send a message to itself through a dedicated shutdown channel saying the shutdown was done. Meanwhile inside `ShardManager::shutdown`, the recv end of the channel was waiting for the message to come in before rebooting the shard.

Reply:
Ok, I was able to simplify the process of booting and restarting shards by a fair bit. Added a `ShardManagerMesssage::Boot` variant that the runner sends after it finishes shutting down. Also, I realized the `PollShardQueue` variant is not strictly needed, and removing it doesn't change control flow nor performance. Maybe you could cause fewer linear scans of the ShardQueue by avoiding checking for a batch every time a new shard is queued, but the queue is small anyways so it doesn't matter.